File: LeetCode/Problems/Python/105. Construct Binary Tree From Preorder and Inorder Traversal.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder or not inorder:
            return None
        
        in_map = dict([(x,i) for i, x in enumerate(inorder)])
        
        head = TreeNode(preorder[0])
        stack = [head]
        for i, x in enumerate(preorder[1:]):
            node = TreeNode(x)
            if in_map[x] < in_map[stack[-1].val]:
                stack[-1].left = node
            else:
                while stack and in_map[x] > in_map[stack[-1].val]:
                    parent = stack.pop()
                parent.right = node
            stack.append(node)
        
        return head

File: LeetCode/Problems/C++/105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(!preorder.size() || !inorder.size()) return NULL;
        
        unordered_map<int, int> in_map;
        for(int i=0;i<inorder.size();i++){
            in_map[inorder[i]] = i;
        }
        
        stack<TreeNode*> mystack;
        TreeNode* head = new TreeNode(preorder[0]);
        mystack.push(head);
        
        for(int i=1;i<preorder.size();i++){
            TreeNode* node = new TreeNode(preorder[i]);
            if(in_map[preorder[i]] < in_map[mystack.top()->val])
                mystack.top()->left = node;
            else{
                TreeNode* parent;
                while(!mystack.empty() && in_map[mystack.top()->val] < in_map[preorder[i]]){
                    parent = mystack.top();
                    mystack.pop();
                }
                parent->right = node;
            }
            mystack.push(node);          
        }
        return head;
        
    }
};